citest/json_predicate/map_predicate.py

```diff
@@ -153,6 +153,21 @@ def __eq__(self, result):
             and self.__good_map == result.good_object_result_mappings
             and self.__bad_map == result.bad_object_result_mappings)
 
+  def clone_with_source(self, source, base_target_path, base_value_path):
+    """Implements CloneableWithNewSource interface."""
+    builder = MapPredicateResultBuilder(self.pred)
+
+    for index, orig in enumerate(self.results):
+      if isinstance(orig, predicate.CloneableWithNewSource):
```

Comment:
Often in python you'd take advantage of duck typing. But I'm using isinstance because I want to indicate class semantics, which I think is appropriate here. The framework has a few marker classes (e.g. "CloneableWithNewSource", "Snapshotable") the guide key heuristics.

CloneableWithNewSource is defined here: https://github.com/google/citest/blob/776b27277dfc74204c5ca649c9f2cde0c62ae0b6/citest/json_predicate/predicate.py#L147

Python classes support multiple inheritence. The base class(es) are specified with the class declaration ("object" being the root class). Here's an example of a predicate result saying that it is CloneableWithNewSource.
https://github.com/google/citest/blob/e85bb27b3bcf1bd2e158c2be1eac0b8964dcf807/citest/json_predicate/path_result.py#L29